src/python/qequlacs/simulator_test.py

```diff
@@ -1,9 +1,178 @@
+            self.assertEqual(max(counts, key=counts.get), "00")
+
+    def test_run_circuit_2qubit_gates(self):
+        # Given
+        circuit = Circuit(Program(X(0), CZ(0, 1)))
+        n_samples = 100
+        # When
+        qulacs_circuit = convert_circuit_to_qulacs(circuit)
+        self.assertEqual(qulacs_circuit.get_gate_count(), 2)
+        self.assertEqual(qulacs_circuit.get_gate(1).get_name(), "CZ")
+
+        for backend in self.backends:
+            backend.n_samples = n_samples
+            measurements = backend.run_circuit_and_measure(circuit)
+
+            counts = measurements.get_counts()
+            self.assertEqual(max(counts, key=counts.get), "10")
+
+    def test_run_circuit_3qubit_gates_toffoli(self):
+        # Given
+        circuit = Circuit(Program(X(0), X(1), CCNOT(0, 1, 2)))
+        n_samples = 100
+        # When
+        qulacs_circuit = convert_circuit_to_qulacs(circuit)
+        self.assertEqual(qulacs_circuit.get_gate_count(), 3)
+        self.assertEqual(qulacs_circuit.get_gate(2).get_name(), "DenseMatrix")
+
+        for backend in self.backends:
+            backend.n_samples = n_samples
+            measurements = backend.run_circuit_and_measure(circuit)
+
+            counts = measurements.get_counts()
+            self.assertEqual(max(counts, key=counts.get), "111")
+
+    def test_run_circuit_3qubit_gates_fredkin(self):
+        # Given
+        circuit = Circuit(Program(X(0), X(2), CSWAP(0, 1, 2)))
+        n_samples = 100
+        # When
+        qulacs_circuit = convert_circuit_to_qulacs(circuit)
+        self.assertEqual(qulacs_circuit.get_gate_count(), 3)
+        self.assertEqual(qulacs_circuit.get_gate(2).get_name(), "DenseMatrix")
+
+        for backend in self.backends:
+            backend.n_samples = n_samples
+            measurements = backend.run_circuit_and_measure(circuit)
+
+            counts = measurements.get_counts()
+            self.assertEqual(max(counts, key=counts.get), "110")
+
+
+class TestGateDefs(unittest.TestCase):
+    """
+    Tests gates XX, YY, ZZ
+    """
+
+    def setUp(self):
+        self.backends = [QulacsSimulator(n_samples=None)]
+        self.wf_simulators = [QulacsSimulator(n_samples=None)]
+        self.qvm = get_qc('9q-square-qvm')
+
+    def test_gate_XX(self):
+        def_gate, gate = XX(0, 0, 1)
+        self.assertEqual(gate.name, "XX")
+        self.assertEqual(len(gate.qubits), 2)
+        self.assertEqual(len(gate.params), 1)
+        self.assertEqual(def_gate.name, "XX")
+
+    def test_append_circuit_XX(self):
+        # Given
+        n_samples = 100
+        # XX with Pi will -> both qubits to 0 in ~100% cases
+        prog1 = Program(X(0), CNOT(0, 1), XX(np.pi, 0, 1))
+        # XX with Pi/2 will -> both qubits to 0 in ~50% cases
+        prog2 = Program(X(0), CNOT(0, 1), XX(np.pi/2, 0, 1))
+        circuit1 = Circuit(prog1)
+        circuit2 = Circuit(prog2)
+        # When
+        results1 = self.qvm.run_and_measure(prog1, trials=10)
```

Comment:
You define `n_samples=100` but the use `trials=10` – seems inconsistent.
Also, having 10 trials is not enough for testing, as tests might sometimes fail due to some edge cases – much better to set it to 100 or even 1000.

Reply:
got it